x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/DeprecationRoleDescriptorConsumer.java

```diff
@@ -58,90 +63,120 @@
 
+    private static final Logger logger = LogManager.getLogger(DeprecationRoleDescriptorConsumer.class);
+
     private final DeprecationLogger deprecationLogger;
     private final ClusterService clusterService;
     private final ThreadPool threadPool;
     private final Set<String> dailyRoleCache;
+    private final Queue<Collection<RoleDescriptor>> workQueue;
+
+    public DeprecationRoleDescriptorConsumer(ClusterService clusterService, ThreadPool threadPool) {
+        this(clusterService, threadPool, new DeprecationLogger(logger));
+    }
 
-    public DeprecationRoleDescriptorConsumer(ClusterService clusterService, ThreadPool threadPool, DeprecationLogger deprecationLogger) {
+    // package-private for testing
+    DeprecationRoleDescriptorConsumer(ClusterService clusterService, ThreadPool threadPool, DeprecationLogger deprecationLogger) {
         this.deprecationLogger = deprecationLogger;
         this.clusterService = clusterService;
         this.threadPool = threadPool;
         // this String Set keeps "<date>-<role>" pairs so that we only log a role once a day.
-        this.dailyRoleCache = Collections.newSetFromMap(Collections.synchronizedMap(new LinkedHashMap<String, Boolean>() {
+        this.dailyRoleCache = Collections.newSetFromMap(new LinkedHashMap<String, Boolean>() {
             @Override
             protected boolean removeEldestEntry(final Map.Entry<String, Boolean> eldest) {
                 return false == eldest.getKey().startsWith(todayISODate());
             }
-        }));
+        });
+        this.workQueue = new LinkedList<>();
     }
 
     @Override
-    public void accept(Collection<RoleDescriptor> effectiveRoleDescriptors) {
-        threadPool.generic().execute(() -> {
-            final SortedMap<String, AliasOrIndex> aliasOrIndexMap = clusterService.state().metaData().getAliasAndIndexLookup();
-            // executing the check asynchronously will not conserve the generated deprecation response headers (which is what we want,
-            // because it's not the request that uses deprecated features, but rather the role definition. Plus, due to caching, we can't
-            // reliably associate response headers to every request).
-            logDeprecatedPermission(effectiveRoleDescriptors, aliasOrIndexMap);
-        });
+    public synchronized void accept(Collection<RoleDescriptor> effectiveRoleDescriptors) {
+        final Collection<RoleDescriptor> uncachedEffectiveRoleDescriptors = new ArrayList<>();
+        for (RoleDescriptor roleDescriptor : effectiveRoleDescriptors) {
+            if (dailyRoleCache.add(buildCacheKey(roleDescriptor))) {
+                uncachedEffectiveRoleDescriptors.add(roleDescriptor);
+            }
+        }
+        if (false == uncachedEffectiveRoleDescriptors.isEmpty()) {
+            workQueue.add(uncachedEffectiveRoleDescriptors);
+            if (workQueue.size() == 1) {
```

Comment:
technically this could double log something. The log thread is already running and pulls the last item off the queue, thread gets suspended after releasing the lock, this method takes the lock and adds a new item, the queue then has a size of 1 so a new runnable is sent to the generic threadpool and the lock gets released, meanwhile the already existing thread wakes up and sees an item on the queue and pulls it off.

I think it might be much simpler if we create a single thread executor using `EsExecutors#newFixed`?

Reply:
I actually believe everything is correct. But it's unorthodox enough that I will go with your suggestion to make the code more maintainable.

> The log thread is already running and pulls the last item off the queue, thread gets suspended after releasing the lock .... the already existing thread wakes up and sees an item on the queue and pulls it off

The pull of the last item **as well as the check** are guarded by the same lock.

The design is unfamiliar because the consumer works on the head item of the work queue, and pops it when he's done with it, as opposed to when the work starts. This is because the non-empty work queue doubles as a flag to the producer, that a consumer is spawned already, so the producer does not have to re-spawn a new consumer, just add the work item to the queue to be processed by the existing consumer.

Reply:
I found it to difficult to create a new `EsExecutors#newFixed` and I also didn't like the idea of having an idle thread sitting around for such an ancillary task as printing deprecation logs.

I have pushed https://github.com/elastic/elasticsearch/pull/38059/commits/e3b3e3232b18a57c33a57c258344b770702da182 that reworks the same original approach by using an extra flag but using the workqueue in a more typical way. I hope this is less confusing? I am okay with either alternatives, and also if you feel strongly we would do better with a dedicated thread, I can work that one out as well.